src/main/java/com/commercetools/sync/services/impl/BaseService.java

```diff
@@ -135,4 +153,156 @@
         }
 
     }
+
+    /**
+     * Given a {@code key}, if it is blank (null/empty), a completed future with an empty optional is returned.
+     * This method then checks if the cached map of resource keys -&gt; ids contains the key. If it does, then an
+     * optional containing the mapped id is returned. If the cache doesn't contain the key; this method attempts to
+     * fetch the id of the key from the CTP project, caches it and returns a
+     * {@link CompletionStage}&lt;{@link Optional}&lt;{@link String}&gt;&gt; in which the result of it's completion
+     * could contain an {@link Optional} with the id inside of it or an empty {@link Optional} if no resource
+     * was found in the CTP project with this key.
+     *
+     * @param key the key by which a resource id should be fetched from the CTP project.
+     * @return {@link CompletionStage}&lt;{@link Optional}&lt;{@link String}&gt;&gt; in which the result of it's
+     *         completion could contain an {@link Optional} with the id inside of it or an empty {@link Optional} if no
+     *         resource was found in the CTP project with this key.
+     */
+    @Nonnull
+    CompletionStage<Optional<String>> fetchCachedResourceId(@Nullable final String key) {
+        if (isBlank(key)) {
+            return CompletableFuture.completedFuture(Optional.empty());
+        }
+        if (keyToIdCache.isEmpty()) {
+            return fetchAndCache(key);
+        }
+        return CompletableFuture.completedFuture(Optional.ofNullable(keyToIdCache.get(key)));
+    }
+
+    /**
+     * Given a {@code key}, this method queries a CTP project and fetches all available resource of given type.
+     * Next it stores key - id pair in the {@code keyToIdCache}.
+     *
+     * @param key the key by which a resource id should be fetched from the CTP project
+     * @return {@link CompletionStage}&lt;{@link Optional}&gt; in which the result of it's completion contains an
+     *         {@link Optional} that contains the matching id if exists, otherwise empty.
+     */
+    abstract CompletionStage<Optional<String>> fetchAndCache(@Nonnull final String key);
```

Comment:
There are few reasons for this:
1. Not all services have caching implemented (`CartDiscountServiceImpl`, `InventoryServiceImpl`)
2. We need a way to provide proper parameters for `#fetchAndCache` method with more parameters (_query_, _keyMapper_ and _typeName_)
3. Most of services have this method implemented already so I thought it would be easier to change it this way
So basically this abstract method should be implemented either as:
`return CompletableFuture.completedFuture(Optional.empty());` if there is no caching needed or `return fetchAndCache(key, ProductQuery::of, Product::getKey, "ProductType");`